echidna/calc/decay.py

```diff
@@ -1,69 +1,113 @@
         self._name = name
-        self._loading = loading
         self._atm_weight_iso = atm_weight_iso
         self._atm_weight_nat = atm_weight_nat
         self._abundance = abundance
         self._phase_space = phase_space
         self._matrix_element = matrix_element
-        self._fv_radius = fv_radius
-        self._scint_density = scint_density
+        self._roi_efficiency = roi_efficiency  # Defaults to standard Gaussian efficiency for -1/2 sigma to +3/2 sigma ROI
+
+    def get_n_atoms(self, fv_radius=None, loading=None, scint_density=None,
+                    target_mass=None, scint_mass=None, outer_radius=None):
+        """ Calculates the number of atoms of the double-beta isotope.
+
+          Set up to follow the full (SNO+-specific) calculation as per
+          SNO+-doc-1728v2 but can look at other scenarios/detectors by
+          overriding the default args.
 
-    def get_n_atoms(self):
-        """ Calculates the number of atoms of an isotope within the
-          fiducial volume.
+        .. warning:: All args default to SNO+ specific values!
+
+        Args:
+          fv_radius (float, optional): Radius of fiducial volume in mm.
+          loading (float, optional): Loading of isotope with 0 to 1
+            equivalent to 0% to 100%.
+          scint_density (float): Density of liquid scintillator in
+            kg/mm^3.
+          target_mass (float, optional): Target mass in kg.
+          scint_mass (float, optional): Mass of scintillator in kg.
+          outer_radius (float, optional): Radius of outer container
+            containing fiducial volume, e.g. AV, in mm.
+
+        Raises:
+          ValueError: If :obj:`loading` is not between zero and 1.
 
         Returns:
           float: Number of atoms.
+
         """
-        if self._loading < 0. or self._loading > 1.:
+        # Set defaults
+        if fv_radius is None:  # use default from constants
+            fv_radius = const._fv_radius
+        if loading is None:  # use default from constants
+            loading = const._loading
+        if loading < 0. or loading > 1.:
             raise ValueError("Loading ranges from 0 to 1")
-        if self._abundance < 0. or self._abundance > 1.:
-            raise ValueError("Abundance ranges from 0 to 1")
-        mass_iso = self._atm_weight_iso*const._atomic_mass_unit  # kg/atom
-        mass_nat = self._atm_weight_nat*const._atomic_mass_unit  # kg/atom
-        mass_fraction = self._abundance*mass_iso/mass_nat
-        if not self._scint_density:
-            self._scint_density = const._scint_density
-        if not self._fv_radius:
-            self._fv_radius = const._fv_radius
-        scint_mass = (4./3.)*numpy.pi*self._fv_radius**3*self._scint_density
-        active_mass = scint_mass*mass_fraction*self._loading
-        return active_mass/mass_iso
-
-    def get_activity(self, half_life, n_atoms):
+        if scint_density is None:  # use default from constants
```

Comment:
I think maybe n_atoms should be a class variable as its used by many functions and only really needs to be calculated once in the initialization